File: test/multiple_inheritance.py
#!/usr/bin/env python
#
#
# Copyright Troy D. Straszheim 2009. Distributed under the Boost
# Software License, Version 1.0. (See accompanying file
# LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
#
#
# Check that multiple inheritance from wrapped classes works.  Note
# that you have to call your base class constructors explicitly in the
# derived class' constructor... or multiple inheritance won't work.
#
#
"""
>>> from multiple_inheritance_ext import *
>>> class Derived(LeftClass, RightClass):
...     def __init__(self):
...         LeftClass.__init__(self)
...         RightClass.__init__(self)
... 
>>> obj = Derived()
>>> test_left(obj)
'Successfully converted to LeftClass'
>>> test_right(obj)
'Successfully converted to RightClass'

    Here's what happens if you don't initialize your base classes.

>>> class BuggyDerived(LeftClass, RightClass):
...     pass
... 
>>> buggy_obj = BuggyDerived()

    This succeeds, but I haven't investigated why.  It'd be nice if it
    didn't.

>>> test_left(buggy_obj)
'Successfully converted to LeftClass'

    Fail as expected

>>> test_right(buggy_obj)
Traceback (most recent call last):
...
Boost.Python.ArgumentError: Python argument types in
    multiple_inheritance_ext.test_right(BuggyDerived)
did not match C++ signature:
    test_right(RightClass)
>>> 
"""

File: test/multiple_inheritance.cpp
#include <boost/python.hpp>
#include <boost/python/detail/logging.hpp>
#include <iostream>

namespace bp = boost::python;

struct LeftClass { };
struct RightClass { };

std::string test_left(const LeftClass &) 
{ 
  return "Successfully converted to LeftClass";
}
std::string test_right(const RightClass &) 
{ 
  return "Successfully converted to RightClass";
}

BOOST_PYTHON_MODULE(multiple_inheritance_ext)
{
    bp::def("test_left", &test_left);
    bp::def("test_right", &test_right);

    bp::class_<LeftClass>("LeftClass", bp::init<>());
    bp::class_<RightClass>("RightClass", bp::init<>());
}
